src/main/java/org/opensearch/ad/transport/ForwardADTaskTransportAction.java

```diff
@@ -26,45 +26,145 @@
+                    adTaskCacheManager.removeRunningEntity(detectorId, adTask.getEntity().get(0).getValue());
+                    if (!adTaskCacheManager.hasEntity(detectorId)) {
+                        logger.info("Historical HC detector done, will remove from cache, detector id:{}", detectorId);
+                        listener.onResponse(new AnomalyDetectorJobResponse(detectorId, 0, 0, 0, RestStatus.OK));
+                        // TODO: reset task state when get task
+                        ADTaskState state = !adTask.isEntityTask() && adTask.getError() != null ? ADTaskState.FAILED : ADTaskState.FINISHED;
+                        adTaskManager.setHCDetectorTaskDone(adTask, state, listener);
+                    } else {
+                        logger.debug("Run next entity for detector " + detectorId);
+                        adTaskManager.runBatchResultActionForEntity(adTask, listener);
+                        adTaskManager
+                            .updateADHCDetectorTask(
+                                detectorId,
+                                adTask.getParentTaskId(),
+                                ImmutableMap
+                                    .of(
+                                        STATE_FIELD,
+                                        ADTaskState.RUNNING.name(),
+                                        TASK_PROGRESS_FIELD,
+                                        adTaskManager.hcDetectorProgress(detectorId),
+                                        ERROR_FIELD,
+                                        adTask.getError() != null ? adTask.getError() : ""
+                                    )
+                            );
+                    }
+                } else {
+                    logger
+                        .warn(
+                            "Can only handle HC entity task for NEXT_ENTITY action, taskId:{} , taskType:{}",
+                            adTask.getTaskId(),
+                            adTask.getTaskType()
+                        );
+                    listener.onFailure(new IllegalArgumentException("Unsupported task"));
+                }
+                break;
+            case PUSH_BACK_ENTITY:
+                // Push back entity to pending entities queue and run next entity.
+                if (adTask.isEntityTask()) { // AD task must be entity level task.
+                    if (adTaskManager.isRetryableError(adTask.getError())
+                        && !adTaskCacheManager.exceedRetryLimit(adTask.getDetectorId(), adTask.getTaskId())) {
+                        // If retryable exception happens when run entity task, will push back entity to the end
+                        // of pending entities queue, then we can retry it later.
+                        adTaskCacheManager.pushBackEntity(adTask.getTaskId(), adTask.getDetectorId(), adTask.getEntity().get(0).getValue());
```

Comment:
Check ADTaskManager line 157 
```
private final Set<String> retryableErrors = ImmutableSet.of(EXCEED_HISTORICAL_ANALYSIS_LIMIT, NO_ELIGIBLE_NODE_TO_RUN_DETECTOR);
```
If task failed to start due to exceed max historical analysis limit or failed to dispatch task due to no eligible node, we will put the entity task to the end of pending entity queue to retry it later. The task state will be set as failed, check `ADBatchTaskRunner` line 522:
```
handleException(adTask, e);
```

Reply:
Checked 
handleException(adTask, e); 
in getDelegatedListener.

Didn't find you set the state to FAILED.

These are the related code I read.
```
e -> {
            listener.onFailure(e);
            handleException(adTask, e);

            if (adTask.isEntityTask()) {
                // if (adTaskManager.isRetryableError(adTask.getError())
                // && !adTaskCacheManager.exceedRetryLimit(adTask.getDetectorId(), adTask.getTaskId())) {
                // // If the error is retryable, move entity from temp queue to the end of pending queue
                // adTaskCacheManager.pushBackEntity(adTask.getTaskId(), adTask.getDetectorId(), adTask.getEntity().get(0).getValue());
                // } else {
                // adTaskCacheManager.removeEntity(adTask.getDetectorId(), adTask.getEntity().get(0).getValue());
                // logger.warn("Entity task failed, task id: {}", adTask.getTaskId());
                // }
                // When reach this line, it means entity task failed to start on worker node
                // Sleep some time before polling next entity task.
                waitBeforeNextEntity(SLEEP_TIME_FOR_NEXT_ENTITY_TASK_IN_MILIS);
                adTaskManager.entityTaskDone(adTask, e, transportService);
                startNewEntityTaskLane(adTask, transportService);
            }
        });
...
private void handleException(ADTask adTask, Exception e) {
        // Check if batch task was cancelled or not by exception type.
        // If it's cancelled, then increase cancelled task count by 1, otherwise increase failure count by 1.
        if (e instanceof ADTaskCancelledException) {
            adStats.getStat(StatNames.AD_CANCELED_BATCH_TASK_COUNT.getName()).increment();
        } else if (ExceptionUtil.countInStats(e)) {
            adStats.getStat(StatNames.AD_BATCH_TASK_FAILURE_COUNT.getName()).increment();
        }
        // Handle AD task exception
        adTaskManager.handleADTaskException(adTask, e);
    }
```

Reply:
Another question: if there is only one entity, how would you handle the retrying case?  Would you quickly exhaust the 3 retries in that case and make the task failed?

Reply:
Need to read more `adTaskManager.handleADTaskException(adTask, e);`
If there is only one, will still sleep some time, check 
```
waitBeforeNextEntity(SLEEP_TIME_FOR_NEXT_ENTITY_TASK_IN_MILIS);
```